### PR TITLE
Add JWT access token support for service accounts

### PR DESCRIPTION
**Pull request by soltanmm**
This is similar to #161, but takes into account the intended use case of simply 'promoting' an assertion to using a self-issued JWT access token instead of [something completely different](https://www.youtube.com/watch?v=K2P86C-1x3o).

But, I'm not quite sure of the appropriate way to test this. There's some test code that covers it (apparently the test code generated the assertion even when the old `create_scoped_required` would have returned `True`), but nothing particularly functional.

It might negatively interact with the spirit of #211.

cc @nathanielmanistaatgoogle @anthmgoogle #252 #253

### CODE REVIEW

File: oauth2client/service_account.py [hunk 1]
```diff
@@ -61,31 +61,50 @@ def __init__(self, service_account_id, service_account_email, private_key_id,
...
+      payload = {
+          'aud': self._token_uri,
+          'scope': self._scopes,
+          'iat': now,
+          'exp': now + _ServiceAccountCredentials.MAX_TOKEN_LIFETIME_SECS,
+          'iss': self._service_account_email
+      }
+      payload.update(self._kwargs)
+
+      first_segment = _urlsafe_b64encode(_json_encode(header))
```
Comment (anthmgoogle) [inline oauth2client/service_account.py:83]: Can more of this code be shared with the copy below.

File: oauth2client/service_account.py [hunk 2]
```diff
@@ -107,9 +126,6 @@ def serialization_data(self):
         'private_key': self._private_key_pkcs8_text
     }
 
-  def create_scoped_required(self):
```
Comment (anthmgoogle) [inline oauth2client/service_account.py:110]: This should be left in for compatibility, although it can be changed to return false.

File: oauth2client/service_account.py [hunk 3]
```diff
@@ -61,31 +61,50 @@ def __init__(self, service_account_id, service_account_email, private_key_id,
   def _generate_assertion(self):
```
Comment (anthmgoogle) [inline oauth2client/service_account.py:61]: This shares the code for the the Jwt creation. However, they are also used differently, as JWT Access goes in directly into the authorization headers as a bearer token, while scoped credentials fetch an access token and use that for authorization token. I was expecting to see some logic reflecting that.

**Comment by anthmgoogle** [comment]
Finished review pass.

File: oauth2client/service_account.py (hunk 3 above)
Comment (soltanmm) [inline oauth2client/service_account.py:61]: I was hoping someone could elucidate for me where that happens in the codebase as part of guidance for writing the tests...

File: oauth2client/service_account.py [hunk 4]
```diff
@@ -61,31 +61,49 @@ def __init__(self, service_account_id, service_account_email, private_key_id,
...
-    first_segment = _urlsafe_b64encode(_json_encode(header))
-    second_segment = _urlsafe_b64encode(_json_encode(payload))
-    assertion_input = first_segment + b'.' + second_segment
-
-    # Sign the assertion.
-    rsa_bytes = rsa.pkcs1.sign(assertion_input, self._private_key, 'SHA-256')
-    signature = base64.urlsafe_b64encode(rsa_bytes).rstrip(b'=')
-
-    return assertion_input + b'.' + signature
+    if self._scopes:
```
Comment (dhermes) [inline oauth2client/service_account.py:76]: What follows is plenty of code added to a method that was already non-trivial. Use helper methods to make testing and comprehension easier.

File: oauth2client/service_account.py [hunk 5]
```diff
@@ -61,31 +61,49 @@ def __init__(self, service_account_id, service_account_email, private_key_id,
...
+      second_segment = _urlsafe_b64encode(_json_encode(payload))
+      assertion_input = first_segment + b'.' + second_segment
+
+      # Sign the assertion.
+      rsa_bytes = rsa.pkcs1.sign(assertion_input, self._private_key, 'SHA-256')
+      signature = base64.urlsafe_b64encode(rsa_bytes).rstrip(b'=')
+
+      return assertion_input + b'.' + signature
+
+    else:
```
Comment (dhermes) [inline oauth2client/service_account.py:93]: So the idea is to enable the new behavior with the existing class. Any reason you don't want to define a sub-class to make the distinction clear?

File: tests/test_service_account.py [hunk 6]
```diff
@@ -76,9 +76,6 @@ def test_service_account_email(self):
     self.assertEqual(self.service_account_email,
                      self.credentials.service_account_email)
 
-  def test_create_scoped_required_without_scopes(self):
-    self.assertTrue(self.credentials.create_scoped_required())
```
Comment (dhermes) [inline tests/test_service_account.py:80]: Why is this deleted?

File: oauth2client/service_account.py [hunk 7]
```diff
@@ -108,7 +126,7 @@ def serialization_data(self):
     }
 
   def create_scoped_required(self):
-    return not self._scopes
+    return False
```
Comment (dhermes) [inline oauth2client/service_account.py:129]: I'm unclear on the reason for this change.

**Comment by dhermes** [comment]
New code paths with no supporting tests. Write the tests please.

Also the design questions is any important one. Why did you elect to do this with the existing class rather than a subclass?

File: oauth2client/service_account.py (hunk 5 above)
Comment (soltanmm) [inline oauth2client/service_account.py:93]: Entirely arbitrary. Earlier I made an attempt to get a clear spec on what it was that needed to be done with respect to this effort, and got unclear responses from multiple sources. That previous attempt used a separate class, but when I was finally informed that the credentials ought to be transparently 'promoted', I scrapped the approach.

File: tests/test_service_account.py (hunk 6 above)
Comment (soltanmm) [inline tests/test_service_account.py:80]: Because `create_scoped_required` arguably became somewhat pointless, and was deleted, and when added back in I forgot to put the test back in (i.e. this in particular is an oversight). I'm still unclear why we'd want to keep that method if it'll always be false and is not meant to be overridden anywhere in the hierarchy (see @anthmgoogle's response in an outdated commit), however, given that I've been told we want an automatic and transparent upgrade of service account credentials in the absence of scopes.

**Comment by soltanmm** [comment]
@dhermes You seem to assume that there was a design to begin with; I'm still trying to figure out what it is that is desired here (and frankly I don't have enough familiarity with the oauth2client library structure or its use to confidently figure out what it is that needs to be here for this feature). At this point I've sort of relegated it to a constraint problem that'll need to be figured out through the pains of circumlocution.

File: oauth2client/service_account.py (hunk 7 above)
Comment (soltanmm) [inline oauth2client/service_account.py:129]: See @anthmgoogle's response in an old commit. I believe I'll have to avoid commit-overwriting for the duration of this process...

File: oauth2client/service_account.py (hunk 4 above)
Comment (soltanmm) [inline oauth2client/service_account.py:76]: I'd understand if this suggestion were about separating the functionality such that said separate functionality may be tested (and I'd love to do that, given, y'know, an actual understanding of what it is that's supposed to be here), but as it stands I don't believe the inherent complexity of this method warrants yet more helper methods s.t. reading it requires jumping around moreso than already necessitated by the current structure.

**Comment by soltanmm** [comment]
@dhermes Actually, can we talk offline? This warrants an offline discussion.

**Comment by dhermes** [comment]
Sorry I forgot to peek at the outdated diffs, that was a fail by me.

I won't have time for a VC until probably Friday, but am happy to chat over email.

File: oauth2client/service_account.py (hunk 5 above)
Comment (dhermes) [inline oauth2client/service_account.py:93]: OK

File: oauth2client/service_account.py (hunk 4 above)
Comment (dhermes) [inline oauth2client/service_account.py:76]: The whole point of a helper is to encapsulate an action into something concrete, e.g. `_make_assertion_for_jwt` and `_use_jwt_as_assertion`. Then each of those concrete actions can be analyzed in isolation.

**Comment by dhermes** [comment]
Got all caught up on the outdated diffs/comments. LMK when / how you want to chat off of GitHub.
